### PR TITLE
[TI-3197] Adjust Booking Timeline in List Item

### PR DESCRIPTION
**Pull request by eDombr**
https://thing-it.atlassian.net/browse/TI-3197

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -72,7 +72,8 @@
     "karma-coverage-istanbul-reporter": "~1.4.2",
     "karma-jasmine": "~1.1.1",
     "karma-jasmine-html-reporter": "^0.2.2",
-    "ng-packagr": "^4.4.0",
+    "ng-packagr": "4.4.0",
```
Comment (Malik19051992) [inline package.json]: is it necessary to use this version?

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<div class="timeline-wrapper" #timeline>
+  <div class="timeline">
+    <div class="division" (click)="book(item)" *ngFor="let item of timelineScale">
+      <span class="time">{{item | moment: 'LT'}}</span>
+    </div>
+    <div class="booked"
+      *ngFor="let b of _blockingBookings"
+      [ngStyle]="{
+        'left': findPositionFromHours(b.start) + 'em',
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html]: I think we can calculate styles ones and set in a private field.

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+})
+export class BookingTimelineComponent implements OnInit, AfterViewInit {
+  @Input() set blockingBookings(bookings: any[]) {
+    this._blockingBookings = bookings.map(b => {
+      const start = moment(b.start);
+      const end = moment(b.end);
+
+      return {
+        ...b,
+        start: start.hours() + start.minutes() / 60,
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: use this logic some times. Move in function.

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+  @Output() bookingChange = new EventEmitter();
+  @Output() bookingDeleted = new EventEmitter();
+
+  @ViewChild('timeline') timeline: ElementRef;
+
+  public _booking: any;
+  public bookingTemp: any;
+  public bookingWidth: number;
+  private bookingWidthTemp: number;
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: order by access modifier: public -> protected -> private

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    this.bookingWidth = (this.bookingTemp.end - this.bookingTemp.start) * 8;
+    this.bookingWidthTemp = this.bookingWidth;
+  }
+
+  @Output() bookingChange = new EventEmitter();
+  @Output() bookingDeleted = new EventEmitter();
+
+  @ViewChild('timeline') timeline: ElementRef;
+
+  public _booking: any;
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: I think it should be private

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  constructor() { }
+
+  public ngOnInit(): void {
+    this.timelineScale = this.fillScale();
+  }
+
+  public ngAfterViewInit(): void {
+    if (this.bookingTemp) {
+      const elem = this.timeline.nativeElement;
+      elem.scrollLeft = this.findPositionFromHours(this.bookingTemp.start) / 0.0625 - 120;
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: magic numbers.

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  public _booking: any;
+  public bookingTemp: any;
+  public bookingWidth: number;
+  private bookingWidthTemp: number;
+
+  public _blockingBookings: any[] = [];
+
+  public timelineScale: Date[];
+  public isBlockCrossing = true;
+  public settings = {
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: add interface for it

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+  public ngAfterViewInit(): void {
+    if (this.bookingTemp) {
+      const elem = this.timeline.nativeElement;
+      elem.scrollLeft = this.findPositionFromHours(this.bookingTemp.start) / 0.0625 - 120;
+    }
+  }
+
+  public onPan(event: any): void {
+    const width = this.bookingWidthTemp + event.deltaX * 0.0625;
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: magic number

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+      if (b.start <= point && point < b.end) {
+        isArea = true;
+      }
+    });
+
+    return isArea;
+  }
+
+  public fillScale(): Date[] {
+    return Array.from({ length: 25 }, (_, i) => i).reduce((r, hour) => {
```
Comment (Malik19051992) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: magic numbers

File: package.json (hunk 1 above)
Comment (eDombr) [reply package.json]: We set it when tried to run the app, just forgot to change it. Fixed

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 3 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: done

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 4 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: done

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 5 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: yep, done

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 6 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: created constants

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 7 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: added

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 8 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: created constants

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 9 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: created constants

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html (hunk 2 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html]: I've removed this function from blocking booking timelines but remained it for current booking cause we will need to recalculate it

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html [hunk 10]
```diff
@@ -0,0 +1,26 @@
+<div class="timeline-wrapper" #timeline>
+  <div class="timeline">
+    <div class="division" (click)="book(item)" *ngFor="let item of timelineScale">
+      <span class="time">{{item | moment: 'LT'}}</span>
+    </div>
+    <div class="booked"
+      *ngFor="let b of _blockingBookings"
+      [ngStyle]="{
+        'left': b.start + 'em',
```
Comment (YurySakovich) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html]: try to use `left.em` and in other places

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 11]
```diff
@@ -0,0 +1,187 @@
+import { Component, OnInit, Input, Output, EventEmitter, ViewChild, ElementRef, AfterViewInit } from '@angular/core';
+
+import * as momentNamespace from 'moment';
+import { Moment } from 'moment';
+
+import { BookingTimelineConfig, BookingTimeline } from './booking-timeline.interfaces';
+const moment = momentNamespace;
```
Comment (YurySakovich) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: for what?

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+@Component({
+  selector: 'ti-booking-timeline',
+  templateUrl: './booking-timeline.component.html',
+  styleUrls: ['./booking-timeline.component.css']
+})
+export class BookingTimelineComponent implements OnInit, AfterViewInit {
+  @Input() set blockingBookings(bookings: any[]) {
+    this._blockingBookings = bookings.map(b => {
```
Comment (YurySakovich) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: may will be better to change b to booking

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+      start: this.convertTime(start),
+      end: this.convertTime(end)
+    };
+
+    this.bookingWidth = this.findPositionFromHours(this.bookingTemp.end - this.bookingTemp.start);
+    this.bookingWidthTemp = this.bookingWidth;
+  }
+  @Output() bookingChange = new EventEmitter();
+  @Output() bookingDeleted = new EventEmitter();
+  @ViewChild('timeline') timeline: ElementRef;
```
Comment (YurySakovich) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: use timelineRef

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+      return;
+    }
+
+    if (this.bookingWidth < this.btConfig.minDivision) {
+      return;
+    }
+    this.bookingWidth = width;
+  }
+
+  public onPanEnd() {
```
Comment (YurySakovich) [inline projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: void

File: projects/thing-it-common-lib/src/lib/thing-it-common-lib.component.ts [hunk 15]
```diff
@@ -1,27 +1,34 @@
...
+    },
+    {
+      _id: 1, start: moment({hours: 2}), end: moment({hours: 3}), canChange: false,
+      bookingUser: { firstName: 'Yury', lastName: 'Sakovich', email: 'yury_sakovich@smartexlab.com' }
+    },
+    {
+      _id: 1, start: moment({hours: 16}), end: moment({hours: 17, minutes: 30}), canChange: false,
+      bookingUser: { firstName: 'Yury', lastName: 'Sakovich', email: 'yury_sakovich@smartexlab.com' }
+    },
+  ];
```
Comment (YurySakovich) [inline projects/thing-it-common-lib/src/lib/thing-it-common-lib.component.ts]: Is it mock data?

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 11 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: It's needed for the building of package

File: projects/thing-it-common-lib/src/lib/thing-it-common-lib.component.ts (hunk 15 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/thing-it-common-lib.component.ts]: yep, I'll remove it

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html (hunk 10 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.html]: you're right, fixed

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 13 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: ok

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 14 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: fixed

File: projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts (hunk 12 above)
Comment (eDombr) [reply projects/thing-it-common-lib/src/lib/booking-timeline/booking-timeline.component.ts]: done

**Pull request merged**